### DISCUSSION TITLE
🎓 OptiPFair v0.4.0 — Knowledge Distillation is here

### DISCUSSION THREAD

**Discussion by peremartra** [peremartra/optipfair#29] [Announcements]
<strong>TL;DR:</strong> You can now prune a model <em>and</em> recover its performance in a single, clean pipeline — all within OptiPFair.</p>
<hr>
## Why this matters

Depth pruning is aggressive. Removing 10 layers from a 24-layer model cuts ~28% of parameters, but it hurts benchmark scores significantly. Until now, recovery required stitching together separate tools, custom training loops, and a lot of guesswork on hyperparameters.

With v0.4.0, the full post-pruning recovery workflow lives inside OptiPFair:

```python
# 1. Analyze which layers matter least (data-driven)
importance_scores = opf.analyze_layer_importance(model, dataloader)

# 2. Remove the weakest layers
student = opf.prune_model_depth(model, layer_indices=layers_to_remove)

# 3. Recover performance via knowledge distillation
trained_student, stats = opf.distill_model(
    student_model=student,
    teacher_model=teacher,
    dataloader=train_dataloader,
    alpha=0.6, beta=0.40,
    temperature=2, skew_alpha=0.3,
    scheduler="cosine", warmup_ratio=0.15,
    epochs=4, learning_rate=2.5e-5,
    return_stats=True,
)
```

The `return_stats=True` flag gives you a full `loss_history` dict for visualizing training curves — no extra bookkeeping needed.

---

## Real numbers

Tested on **Qwen3.5-0.8B-Base** (A100, Cosmopedia dataset, 40K samples):

| Metric | Teacher | Student (KD) |
|--------|---------|--------------|
| Layers | 24 | 14 |
| Parameters | 752M | 540M |
| Param Reduction | — | **28.2%** |
| Arc Easy | 67.5% | 50.0% |
| Winogrande | 59.4% | 54.8% |
| Hellaswag | 54.9% | 38.5% |
| Lambada OpenAI | 50.9% | 36.5% |
| PIQA | 71.5% | 64.5% |
| **Avg Benchmark** | **60.8%** | **48.9%** |

Also validated on **Llama-3.2-3B** with equivalent results.

---

## Two notebooks to get started

We ship two versions depending on your use case:

- **[[Knowledge Distillation](https://github.com/peremartra/optipfair/blob/main/examples/knowledge_distillation.ipynb)]** — Full workflow: prune → distill → push to HF → benchmark with `lm_eval` → VRAM/throughput comparison. Recommended if you want end-to-end validation.
- **[[Knowledge Distillation Express](https://github.com/peremartra/optipfair/blob/main/examples/knowledge_distillation_express.ipynb)]** — Compact version focused on the distillation loop with `lm_eval` benchmarks. Good for quick experimentation.

Both run on Google Colab with an A100 (T4 works for smaller models).

---

## What's configurable in `opf.distill_model()`

The API exposes the full distillation setup without requiring you to write a training loop:

| Parameter | What it controls |
|-----------|-----------------|
| `alpha` / `beta` | Weight of task loss (cross-entropy) vs logits loss (Skew KLD) |
| `gamma` / `delta` | Feature alignment losses — set to 0 for labels-only distillation |
| `temperature` | Softens teacher distribution for knowledge transfer |
| `skew_alpha` | Controls asymmetry in the KL divergence |
| `scheduler` | `"cosine"` or `"linear"` with `warmup_ratio` |
| `layer_mapping_strategy` | How teacher layers map to student during alignment |

If you want your LLM assistant to help you tune these, drop [`[optipfair_llm_reference_manual.txt](https://github.com/peremartra/optipfair)`](https://github.com/peremartra/optipfair) into its context — it has everything needed to reason about the parameters correctly.

---

## What's next

v0.4.0 focuses on depth pruning recovery. Width pruning recovery and attention-aware distillation are on the roadmap.

Questions, results, or issues — drop them here or open an issue. If the notebooks save you time, a ⭐ on the repo goes a long way.
